### ISSUE TITLE
Is the "epsilon" parameter of the MultiSimilarityMiner working as expected?

### ISSUE THREAD

**Issue by NateThom** [KevinMusgrave/pytorch-metric-learning#691]
I apologize in advance if this is a nonsensical issue. I fear that it might be as nobody else seems to have an issue with it. Regardless, I'll ask just in case.

I am working with the MultiSimilarityMiner to select hard negative pairs for Contrastive Learning. When attempting to tune hyperparameters of the model I noted that aggressive changes to the epsilon did not reduce the number of hard pairs to always be none. The documentation states the following about the epsilon parameter: "Positive pairs are chosen if they have similarity less than the hardest negative pair, plus this margin (epsilon). " For me, this reads to say that increases to the epsilon parameter will result in pairs which are an "epsilon" less similar than the hardest negative pair. The documentation also states that the default distance function for MultiSimilarityMiner is CosineSimilarity.

I debugged the issue into pytorch-metric-learning source and found the code block which is responsible for identifying the the hard mined samples in "multi-similarity-miner.py", lines 43-56.
```python
if self.distance.is_inverted:
    hard_pos_idx = torch.where(
        pos_sorted - self.epsilon < neg_sorted[:, -1].unsqueeze(1)
    )
    hard_neg_idx = torch.where(
        neg_sorted + self.epsilon > pos_sorted[:, 0].unsqueeze(1)
    )
else:
    hard_pos_idx = torch.where(
        pos_sorted + self.epsilon > neg_sorted[:, 0].unsqueeze(1)
    )
    hard_neg_idx = torch.where(
        neg_sorted - self.epsilon < pos_sorted[:, -1].unsqueeze(1)
    )
```

According to my interpretation of the docs (increases to epsilon result in harder mined pairs), the expected condition would be in the else portion of this if-else statement.

"self.distance" is set by default to CosineSimilarity, which inherits from the BaseDistance class. First, the CosineSimilarity object is required have the is_inverted flag set to True (which makes sense based on the operation of class member functions like "smallest_dist". This means that this conditional statement might not be necessary, but I am all for redundant error checking so no problem here. Second (and more importantly), we will always drop into the first condition of this if-else statement. In english, the interpretation of the torch.where call is "select any positive sample which is less similar than the least similar negative sample, minus epsilon". This is causing increases to the epsilon parameter to select easier samples rather than harder samples.

I don't have a dog in the fight about whether or not epsilon should be making the mined samples more or less hard. I will also happily admit defeat if I have done this analysis wrong. Just trying to be helpful!

Epsilon set to 10
![Screenshot 2024-03-25 at 8 58 30 PM](https://github.com/KevinMusgrave/pytorch-metric-learning/assets/39501007/ff15dc4a-4333-4524-9bc8-24c5783d7355)

The first value in this row of pos_sorted is a candidate value (not set to infinity)
![Screenshot 2024-03-25 at 8 59 03 PM](https://github.com/KevinMusgrave/pytorch-metric-learning/assets/39501007/9779a597-9362-433a-a439-9caadb33084d)

After subtracting epsilon, the value becomes less similar than any negative sample could possibly be (cosine similarity produces values between -1 and 1).
![Screenshot 2024-03-25 at 9 00 51 PM](https://github.com/KevinMusgrave/pytorch-metric-learning/assets/39501007/9db3365e-ae61-49e4-be38-148c97d67332)

The cosine similarity of the corresponding negative sample
![Screenshot 2024-03-25 at 9 07 12 PM](https://github.com/KevinMusgrave/pytorch-metric-learning/assets/39501007/dfce630e-da13-4ff7-9ee8-e5c5fcae1b98)

The operation results in this sample being selected as a hard positive, despite an epsilon of 10. I believe the correct functionality would be for this hard positive to be filtered out.
![Screenshot 2024-03-25 at 9 01 45 PM](https://github.com/KevinMusgrave/pytorch-metric-learning/assets/39501007/c1208ac8-4654-43be-ab20-3c6a236399a1)

**Comment by KevinMusgrave** [comment]
This may be an issue with the wording in the documentation. 

Here are the equations from the original paper.

Negative pairs:
![image](https://github.com/KevinMusgrave/pytorch-metric-learning/assets/2314891/261fad0e-3d07-432b-acc8-085a446622cc)

Positive pairs:
![image](https://github.com/KevinMusgrave/pytorch-metric-learning/assets/2314891/830039a4-f192-49f1-9d53-22126e40505a)

In the above equations, "S" is cosine similarity.

**Comment by NateThom** [comment]
Thanks Kevin! You're a legend for responding to this. I know it was a lot.
